flang/lib/Optimizer/CodeGen/CodeGen.cpp

```diff
@@ -143,36 +144,39 @@ class FIRToLLVMTypeConverter : public mlir::LLVMTypeConverter {
     return mlir::LLVM::LLVMType::getInt64Ty(llvmDialect);
   }
 
+  // TODO
+  bool requiresExtendedDesc() { return false; }
+
   // This corresponds to the descriptor as defined ISO_Fortran_binding.h and the
   // addendum defined in descriptor.h.
-  // FIXME: This code should be generated and follow SPOT
-  mlir::LLVM::LLVMType convertBoxType(fir::BoxType box) {
+  mlir::LLVM::LLVMType convertBoxType(fir::BoxType box, int rank = -1) {
```

Comment:
The internals of this functions seems very dependent on the descriptor format (order of fields...), could part of this logic be moved to DescriptorModel.h (taking the element type and rank as argument) ?

Reply:
It's a chicken-and-egg problem, for sure.  There is no declarative description of a descriptor. It's spread across multiple header files in possibly two source languages.  C++'s support for reflection in these cases is uneven at best (or could just be argued to be totally broken).

There is no getting around that we have to be able to generate a descriptor, whatever it happens to be. While we can make some of that machinery automajik on the headers, there is a certain amount of information that just has to be taught directly to Tilikum.